### PR TITLE
Implementation Plan: Split monolithic file recipe/rules/rules_campaign.py (1076 lines)

### LINKED ISSUE #2842: Split monolithic file: recipe/rules/rules_campaign.py (1076 lines)
Split from #2832 — high-effort standalone file.

## File

| File | Lines |
|------|-------|
| `recipe/rules/rules_campaign.py` | 1076 |

## Context

Largest file in the monolithic split wave. At 1076 lines this is high effort — requires identifying extraction targets, creating new submodules, rewiring imports across callers.

Part of the P8 finding from the arch audit. Prior split wave (#1390, #1392) completed; this file has accumulated since.

### PR DESCRIPTION
**Pull request by Trecek**
## Summary

Split `rules_campaign.py` (1076 lines, 29 `@semantic_rule` functions + 6 helpers/constants blocks) into 5 focused submodules by thematic cohesion. Extract 2 shared helpers (`_load_dispatch_target`, `_extract_sentinel_fields`) to `_rule_helpers.py` — the canonical shared-helper module for recipe rules — to comply with the "no cross-imports between rule modules" architectural invariant. Delete the original file and update all registry/import consumers.

**New files:**
| File | Rules | ~Lines |
|------|-------|--------|
| `rules_campaign_dispatch.py` | 9 (structure, naming, recipe refs, task) | ~250 |
| `rules_campaign_ingredients.py` | 4 (ingredient keys, dangling, required, types) | ~142 |
| `rules_campaign_deps.py` | 3 (depends_on valid, acyclic, sequential) | ~99 |
| `rules_campaign_capture.py` | 4 (capture keys, values, sentinel field, all sentinels) | ~184 |
| `rules_campaign_flow.py` | 9 (gates, paths, campaign refs, version, skip-when) | ~348 |

**Deleted:** `rules_campaign.py`

Closes #2842

## Implementation Plan

Plan file: `/home/talon/projects/autoskillit-runs/impl-20260523-231344-931172/.autoskillit/temp/make-plan/split_rules_campaign_plan_2026-05-23_232600.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code) via AutoSkillit

## Token Usage Summary

| Step | Model | count | uncached | output | cache_read | peak_ctx | turns | cache_write | time |
|------|-------|-------|----------|--------|------------|----------|-------|-------------|------|
| plan | claude-sonnet-4-6 | 1 | 99 | 24.5k | 2.8M | 116.7k | 197 | 95.2k | 21m 38s |
| verify | claude-opus-4-6 | 1 | 58 | 12.9k | 1.9M | 84.7k | 135 | 69.4k | 7m 29s |
| implement* | MiniMax-M2.7-highspeed | 1 | 4.9M | 47.1k | 2.9M | 95.0k | 185 | 77.7k | 15m 45s |
| prepare_pr* | MiniMax-M2.7-highspeed | 1 | 164.2k | 5.5k | 356.7k | 30.0k | 29 | 15.0k | 2m 7s |
| compose_pr* | MiniMax-M2.7-highspeed | 1 | 55.8k | 1.9k | 236.8k | 30.0k | 16 | 14.7k | 55s |
| review_pr | claude-sonnet-4-6 | 2 | 602 | 59.9k | 2.7M | 81.3k | 192 | 146.8k | 22m 1s |
| resolve_review | claude-sonnet-4-6 | 2 | 118 | 39.9k | 5.7M | 147.0k | 202 | 178.6k | 20m 21s |
| ci_conflict_fix | claude-opus-4-6 | 2 | 80 | 9.3k | 1.8M | 55.1k | 83 | 84.1k | 4m 53s |
| resolve_queue_merge_conflicts | claude-opus-4-6 | 2 | 119 | 19.4k | 4.2M | 69.4k | 152 | 96.2k | 9m 32s |
| **Total** | | | 5.1M | 220.3k | 22.5M | 147.0k | | 777.9k | 1h 44m |

\* *Step used a non-Anthropic provider; caching behavior may differ.*

## Token Efficiency

| Step | LoC Changed | cache_read/LoC | cache_write/LoC | output/LoC |
|------|-------------|----------------|-----------------|------------|
| plan | 0 | — | — | — |
| verify | 0 | — | — | — |
| implement | 2288 | 1247.4 | 34.0 | 20.6 |
| prepare_pr | 0 | — | — | — |
| compose_pr | 0 | — | — | — |
| review_pr | 0 | — | — | — |
| resolve_review | 0 | — | — | — |
| ci_conflict_fix | 1289 | 1390.9 | 65.3 | 7.2 |
| resolve_queue_merge_conflicts | 6276 | 662.4 | 15.3 | 3.1 |
| **Total** | **9853** | 2283.7 | 79.0 | 22.4 |

## Model Usage Breakdown

| Model | steps | uncached | output | cache_read | cache_write | time |
|-------|-------|----------|--------|------------|-------------|------|
| claude-sonnet-4-6 | 3 | 819 | 124.3k | 11.2M | 420.7k | 1h 4m |
| claude-opus-4-6 | 3 | 257 | 41.5k | 7.9M | 249.7k | 21m 54s |
| MiniMax-M2.7-highspeed | 3 | 5.1M | 54.5k | 3.4M | 107.5k | 18m 47s |

### CODE REVIEW

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review — Verdict: changes_requested

File: tests/docs/test_doc_counts.py [hunk 1]
```diff
@@ -342,7 +342,7 @@ def test_orchestration_states_13_retry_reasons() -> None:
     _assert_doc_states_number(DOCS_DIR / "execution" / "orchestration.md", "retry reasons", 13)
 
 
-def test_authoring_states_24_rule_families() -> None:
+def test_authoring_states_28_rule_families() -> None:
     _assert_doc_states_number(DOCS_DIR / "recipes" / "authoring.md", "rule families", 24)
```
Comment (Trecek) [inline tests/docs/test_doc_counts.py:346]: [critical] tests: `test_authoring_states_28_rule_families` was renamed (24→28 in the function name) but `_assert_doc_states_number(..., "rule families", 24)` still asserts the old value 24. `authoring.md` was updated to state 28 rule families (line L83), so this test will fail at runtime. Change the expected count argument from 24 to 28.

File: src/autoskillit/recipe/_rule_helpers.py [hunk 2]
```diff
@@ -55,6 +63,40 @@ def _is_failure_sentinel_value(val: Any) -> bool:
     return val is False or (isinstance(val, str) and val.lower() == "false")
 
 
+def _extract_sentinel_fields(recipe: Recipe) -> frozenset[str]:
+    """Extract declared field names from all sentinel stop step JSON examples."""
+    fields: set[str] = set()
+    for step in recipe.steps.values():
+        if step.action != "stop" or not step.message:
+            continue
+        if "sentinel" not in step.message.lower():
```
Comment (Trecek) [inline src/autoskillit/recipe/_rule_helpers.py:72]: [warning] defense: Sentinel heuristic `"sentinel" not in step.message.lower()` may miss stop steps that use alternative phrasing. `_SENTINEL_TRIGGER_RE` (defined at L20–L23 in this module) already covers multiple phrases — use it here instead of the bare string check to eliminate false negatives from `_extract_sentinel_fields`.

File: src/autoskillit/recipe/_rule_helpers.py [hunk 3]
```diff
@@ -55,6 +63,40 @@ def _is_failure_sentinel_value(val: Any) -> bool:
...
+    if project_dir is None:
+        return None
+    try:
+        from autoskillit.recipe.io import find_recipe_by_name, load_recipe  # noqa: PLC0415
+
+        info = find_recipe_by_name(dispatch.recipe, project_dir)
+        if info is None:
+            return None
+        return load_recipe(info.path)
+    except Exception:
```
Comment (Trecek) [inline src/autoskillit/recipe/_rule_helpers.py:95]: [warning] defense: `_load_dispatch_target` catches bare `Exception`, collapsing 'recipe not found' and 'recipe parse error' into the same `None` return. Callers across all five campaign submodules cannot distinguish expected skip from unexpected failure. Narrow to `FileNotFoundError`/`ValueError` and re-raise (or log at ERROR) for unexpected exceptions.

File: src/autoskillit/recipe/rules/rules_campaign_capture.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    Returns a list of frozensets, one per success sentinel stop, rather than
+    the union. This enables per-path validation. Failure-terminal sentinels
+    (those whose example JSON contains ``"success": false``) are excluded
+    because failure paths do not produce captured result fields.
+    """
+    per_stop: list[frozenset[str]] = []
+    for step in recipe.steps.values():
+        if step.action != "stop" or not step.message:
+            continue
+        if "sentinel" not in step.message.lower():
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_campaign_capture.py:96]: [warning] defense: `_extract_sentinel_fields_per_stop` duplicates the fragile `"sentinel" not in step.message.lower()` heuristic from `_rule_helpers._extract_sentinel_fields`. If the heuristic is widened or replaced with `_SENTINEL_TRIGGER_RE` in the helper, this copy must be updated separately. Unify the guard to prevent divergence.

File: src/autoskillit/recipe/rules/rules_campaign_deps.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    def dfs(node: str, path: list[str]) -> None:
+        visited.add(node)
+        in_stack.add(node)
+        for neighbor in adjacency.get(node, []):
+            if neighbor not in adjacency:
+                continue
+            if neighbor not in visited:
+                dfs(neighbor, path + [neighbor])
+            elif neighbor in in_stack:
+                cycle_start = path.index(neighbor) if neighbor in path else 0
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_campaign_deps.py:60]: [warning] bugs: Cycle path reconstruction fallback `else 0` is both unreachable and incorrect. When `neighbor in in_stack` is True, `neighbor` must be in `path` (DFS invariant — path and in_stack are kept in sync). The fallback would report the cycle starting from `path[0]` (DFS root) instead of the actual cycle entry point, producing a misleading error message. Remove the `else 0` branch or replace with `assert neighbor in path`.

File: docs/recipes/authoring.md [hunk 6]
```diff
@@ -47,14 +47,18 @@ top-level keys (defined in `recipe/schema.py:Recipe`):
 ## The 24 semantic rule families
 
 `validate_recipe` runs every rule registered with `recipe/registry.py:semantic_rule`.
-The rule families live in `src/autoskillit/recipe/rules_*.py` (24 files):
+The rule families live in `src/autoskillit/recipe/rules_*.py` (28 files):
```
Comment (Trecek) [inline docs/recipes/authoring.md:50]: [warning] cohesion: Doc claims '28 files' for rule families but `test_semantic_rule_family_count_is_36` in `test_doc_counts.py` enforces 36. If 36 is the correct total file count, update the doc to '36 files'. If 28 is a curated-family count distinct from the file count, add a clarifying note and update the section heading (L47 still reads '## The 24 semantic rule families').

File: docs/recipes/authoring.md [hunk 7]
```diff
@@ -47,14 +47,18 @@ top-level keys (defined in `recipe/schema.py:Recipe`):
 ## The 24 semantic rule families
```
Comment (Trecek) [inline docs/recipes/authoring.md:47]: [warning] cohesion: Section heading '## The 24 semantic rule families' is stale — the body text was updated to 28 but the heading was not changed. Update the heading to match the body (28) or the enforced file count (36).

File: src/autoskillit/recipe/__init__.py [hunk 8]
```diff
@@ -101,7 +101,17 @@
 from autoskillit.recipe.rules import (  # noqa: E402
     rules_callable_scope as _rules_callable_scope,  # noqa: F401
 )
-from autoskillit.recipe.rules import rules_campaign as _rules_campaign  # noqa: E402 F401
+from autoskillit.recipe.rules import (  # noqa: E402
```
Comment (Trecek) [inline src/autoskillit/recipe/__init__.py:104]: [warning] cohesion: The 5 new campaign submodule imports use inconsistent style — lines L104–L106, L108–L110, and L112–L114 use parenthesized multi-line form while lines L107 and L111 use the single-line form. All surrounding rule imports (rules_ci, rules_clone, etc.) uniformly use the single-line form. Normalize all 5 new additions to the single-line pattern.

File: src/autoskillit/recipe/rules/CLAUDE.md [hunk 9]
```diff
@@ -1,6 +1,6 @@
 # rules/
 
-Semantic validation rule modules for recipe analysis (31 rule files).
+Semantic validation rule modules for recipe analysis (35 rule files).
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/CLAUDE.md]: [warning] cohesion: `CLAUDE.md` states 35 rule files but `test_subpackage_isolation.py` enforces 37 (see L697 hunk). Update the count in this file to 37 to match the enforced value.

File: tests/recipe/test_rules_campaign.py [hunk 10]
```diff
@@ -1399,7 +1399,7 @@ def test_extract_sentinel_fields_handles_nested_braces(tmp_path: Path):
     invalid JSON that json.loads() fails on. This should be silently caught,
     leaving sentinel_fields empty — which is the bug.
     """
-    from autoskillit.recipe.rules.rules_campaign import _extract_sentinel_fields
+    from autoskillit.recipe._rule_helpers import _extract_sentinel_fields
```
Comment (Trecek) [inline tests/recipe/test_rules_campaign.py:1402]: [warning] tests: `TestCampaignRulesCapture` docstring says tests were 'moved to test_campaign_loader.py' but the class body still contains test methods. If tests were truly moved, remove the class body or the class entirely to prevent duplicate coverage and confusion about the canonical test location.

File: src/autoskillit/recipe/_rule_helpers.py [hunk 11]
```diff
@@ -55,6 +63,40 @@ def _is_failure_sentinel_value(val: Any) -> bool:
...
+            try:
+                parsed = json.loads(block)
+                if isinstance(parsed, dict):
+                    fields.update(parsed.keys())
+            except (json.JSONDecodeError, ValueError):
+                continue
+    return frozenset(fields)
+
+
+def _load_dispatch_target(dispatch: CampaignDispatch, project_dir: Path | None) -> Recipe | None:
```
Comment (Trecek) [inline src/autoskillit/recipe/_rule_helpers.py:84]: [warning] tests: `_load_dispatch_target` was promoted from a private function in `rules_campaign.py` to a shared helper in `_rule_helpers.py` used by all five campaign submodules, but has no direct unit tests. Its three distinct None-return paths (project_dir is None, recipe not found, exception) are unexercised at the helper level.

File: src/autoskillit/recipe/rules/rules_campaign_flow.py [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+logger = get_logger(__name__)
+
+_VALID_GATE_TYPES: frozenset[DispatchGateType] = frozenset({DispatchGateType.CONFIRM})
+
+_INPUTS_REF_RE = re.compile(r"\$\{\{\s*inputs\.(\w+)\s*\}\}")
+_ANY_TEMPLATE_REF_RE = re.compile(r"\$\{\{\s*(?:inputs|campaign)\.\w+\s*\}\}")
+_SKIP_EXPR_RE = re.compile(r"^.+\s+(?:==|!=)\s+.+$")
+
+
+def _build_ancestors(name: str, adjacency: dict[str, list[str]]) -> set[str]:
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_campaign_flow.py:27]: [warning] tests: `_build_ancestors` was moved from `rules_campaign.py` to `rules_campaign_flow.py` with no direct unit tests from the new location. Add tests verifying transitive ancestor closure (e.g., A→B→C correctly yields all 3 as ancestors of C).

File: src/autoskillit/recipe/rules/rules_campaign_dispatch.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+
+import regex as re
+
+from autoskillit.core import RECIPE_PACK_REGISTRY, Severity, get_logger
+from autoskillit.recipe._analysis import ValidationContext
+from autoskillit.recipe._rule_helpers import _load_dispatch_target
+from autoskillit.recipe.registry import RuleFinding, semantic_rule
+from autoskillit.recipe.schema import RecipeKind
+
+if TYPE_CHECKING:
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_campaign_dispatch.py:16]: [warning] slop: Empty `if TYPE_CHECKING: pass` block imports nothing and should be removed. The same dead pattern appears in `rules_campaign_flow.py` (L15) and `rules_campaign_ingredients.py` (L13).

File: src/autoskillit/recipe/rules/rules_campaign_flow.py [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+
+import regex as re
+
+from autoskillit.core import DispatchGateType, Severity, get_logger
+from autoskillit.recipe._analysis import ValidationContext
+from autoskillit.recipe._rule_helpers import _load_dispatch_target
+from autoskillit.recipe.registry import RuleFinding, semantic_rule
+from autoskillit.recipe.schema import CAMPAIGN_REF_RE, RecipeKind
+
+if TYPE_CHECKING:
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_campaign_flow.py:15]: [warning] slop: Empty `if TYPE_CHECKING: pass` block imports nothing and should be removed.

File: src/autoskillit/recipe/rules/rules_campaign_ingredients.py [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+
+from typing import TYPE_CHECKING
+
+from autoskillit.core import Severity
+from autoskillit.recipe._analysis import ValidationContext
+from autoskillit.recipe._rule_helpers import _load_dispatch_target
+from autoskillit.recipe.registry import RuleFinding, semantic_rule
+from autoskillit.recipe.schema import RecipeKind
+
+if TYPE_CHECKING:
```
Comment (Trecek) [inline src/autoskillit/recipe/rules/rules_campaign_ingredients.py:13]: [warning] slop: Empty `if TYPE_CHECKING: pass` block imports nothing and should be removed.

File: tests/docs/test_doc_counts.py [hunk 16]
```diff
@@ -271,8 +271,8 @@ def test_retry_reason_value_count_is_14() -> None:
     assert len(values) == 14, f"RetryReason has {len(values)} values: {values}"
 
 
-def test_semantic_rule_family_count_is_32() -> None:
-    assert _count_semantic_rule_files() == 32
+def test_semantic_rule_family_count_is_36() -> None:
```
Comment (Trecek) [inline tests/docs/test_doc_counts.py]: [warning] tests: `test_semantic_rule_family_count_is_36` enforces 36 but `authoring.md` now states '28 rule families'. Clarify whether the test counts all `rules_*.py` files (36) while the doc lists a curated subset (28). If they measure different things, document the distinction; if one is wrong, fix it.

**Review by Trecek — COMMENTED** [review]
AutoSkillit review: 1 critical finding (test assertion mismatch in test_doc_counts.py:346) and 15 warnings detected. See inline comments — blocking issue requires resolution before merge.

File: tests/docs/test_doc_counts.py (hunk 1 above)
Comment (Trecek) [reply tests/docs/test_doc_counts.py:346]: Agreed — fixed in 6f000caf. Updated test function name to test_authoring_states_36_rule_families and assertion from 24 to 36 to match actual rules_*.py file count.

File: src/autoskillit/recipe/_rule_helpers.py (hunk 2 above)
Comment (Trecek) [reply src/autoskillit/recipe/_rule_helpers.py:72]: Agreed — fixed in e9d36223. Replaced bare string check with _SENTINEL_TRIGGER_RE.search(step.message) in both _extract_sentinel_fields and _extract_sentinel_fields_per_stop.

File: src/autoskillit/recipe/_rule_helpers.py (hunk 3 above)
Comment (Trecek) [reply src/autoskillit/recipe/_rule_helpers.py:95]: Agreed — fixed in e9d36223. Narrowed to except (FileNotFoundError, ValueError, OSError) so unexpected failures propagate.

File: src/autoskillit/recipe/rules/rules_campaign_capture.py (hunk 4 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_campaign_capture.py:96]: Agreed — fixed in e9d36223. Imported _SENTINEL_TRIGGER_RE from _rule_helpers and replaced the duplicate string check in _extract_sentinel_fields_per_stop.

File: src/autoskillit/recipe/rules/rules_campaign_deps.py (hunk 5 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_campaign_deps.py:60]: Agreed — fixed in fad7b099. Removed the unreachable else 0 branch; the DFS invariant guarantees neighbor is in path when neighbor is in in_stack.

File: docs/recipes/authoring.md (hunk 6 above)
Comment (Trecek) [reply docs/recipes/authoring.md:50]: Agreed — fixed in 6f000caf. Updated doc body from 28 to 36, added 8 missing table entries (rules_callable_scope, rules_food_truck, rules_inline_script, rules_merge_queue, rules_optional_capture, rules_remediation, rules_route_gate, rules_loop_progress).

File: docs/recipes/authoring.md (hunk 7 above)
Comment (Trecek) [reply docs/recipes/authoring.md:47]: Agreed — fixed in 6f000caf. Heading updated from '## The 24 semantic rule families' to '## The 36 semantic rule families'.

File: src/autoskillit/recipe/__init__.py (hunk 8 above)
Comment (Trecek) [reply src/autoskillit/recipe/__init__.py:104]: Investigated — this is intentional. The inconsistency is driven by ruff's line-length enforcement (88-char limit). Lines like `rules_campaign_capture` and `rules_campaign_ingredients` exceed 88 chars as single-line imports, so ruff auto-formats them to parenthesized form. The pre-existing `rules_callable_scope` import (L101-103) uses the same parenthesized pattern for the same reason. Normalizing all to single-line would cause ruff to re-split them on the next format pass.

File: src/autoskillit/recipe/rules/CLAUDE.md (hunk 9 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/CLAUDE.md]: Agreed — fixed in 6f000caf. Updated count from 35 to 37 rule files.

File: tests/recipe/test_rules_campaign.py (hunk 10 above)
Comment (Trecek) [reply tests/recipe/test_rules_campaign.py:1402]: Valid observation — flagged for design decision. The class body does contain active test methods while a docstring mentions tests were moved. Requires human decision on whether to remove the class, update the docstring, or keep both locations for different coverage purposes.

File: src/autoskillit/recipe/_rule_helpers.py (hunk 11 above)
Comment (Trecek) [reply src/autoskillit/recipe/_rule_helpers.py:84]: Valid observation — flagged for design decision. Adding direct unit tests for _load_dispatch_target's three None-return paths is a scope decision for the implementation PR. The function is exercised indirectly through campaign rule tests.

File: src/autoskillit/recipe/rules/rules_campaign_flow.py (hunk 12 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_campaign_flow.py:27]: Valid observation — flagged for design decision. Adding direct unit tests for _build_ancestors transitive closure is a scope decision. The function is covered indirectly by campaign-ingredient-refs-have-prior-capture rule tests.

File: src/autoskillit/recipe/rules/rules_campaign_dispatch.py (hunk 13 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_campaign_dispatch.py:16]: Agreed — fixed in d6455a36. Removed empty if TYPE_CHECKING: pass block and unused TYPE_CHECKING import.

File: src/autoskillit/recipe/rules/rules_campaign_flow.py (hunk 14 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_campaign_flow.py:15]: Agreed — fixed in d6455a36. Removed empty if TYPE_CHECKING: pass block and unused TYPE_CHECKING import.

File: src/autoskillit/recipe/rules/rules_campaign_ingredients.py (hunk 15 above)
Comment (Trecek) [reply src/autoskillit/recipe/rules/rules_campaign_ingredients.py:13]: Agreed — fixed in d6455a36. Removed empty if TYPE_CHECKING: pass block and unused TYPE_CHECKING import.

File: tests/docs/test_doc_counts.py (hunk 16 above)
Comment (Trecek) [reply tests/docs/test_doc_counts.py]: Agreed — fixed in 6f000caf. Both the doc (authoring.md) and the test now use 36 as the canonical rule family count. The test counts all rules_*.py files (36), and the doc table now lists all 36 entries.
